packages/@sanity/form-builder/src/inputs/PortableText/BlockActions.tsx

```diff
@@ -0,0 +1,75 @@
+import {
+  PortableTextEditor,
+  PortableTextBlock,
+  usePortableTextEditor,
+} from '@sanity/portable-text-editor'
+import React, {useCallback, useMemo} from 'react'
+import PatchEvent from '../../../PatchEvent'
+import {createBlockActionPatchFn} from './utils/createBlockActionPatchFn'
+import {RenderBlockActions} from './types'
+
+type BlockActionsProps = {
+  block: PortableTextBlock
+  onChange: (event: PatchEvent) => void
+  renderBlockActions?: RenderBlockActions
+}
+
+const noSelectStyle: React.CSSProperties = {userSelect: 'none', display: 'flex'}
+
+function isClassComponent(component) {
+  return typeof component === 'function' && !!component.prototype?.isReactComponent
+}
+
+function isFunctionComponent(component) {
+  return typeof component === 'function' && String(component).includes('return React.createElement')
+}
+
+export function BlockActions(props: BlockActionsProps) {
+  const editor = usePortableTextEditor()
+  const {block, onChange, renderBlockActions} = props
+  const decoratorValues = useMemo(
+    () => PortableTextEditor.getPortableTextFeatures(editor).decorators.map((d) => d.value),
+    [editor]
+  )
+
+  const blockActions = useMemo(() => {
+    if (renderBlockActions) {
+      const blockActionProps = {
+        block,
+        value: PortableTextEditor.getValue(editor),
+        set: createBlockActionPatchFn('set', block, onChange, decoratorValues),
+        unset: createBlockActionPatchFn('unset', block, onChange, decoratorValues) as () => void,
+        insert: createBlockActionPatchFn('insert', block, onChange, decoratorValues),
+      }
+      // Support returning a class component for renderBlockActions (to keep backward compatability as it was possible before)
+      if (isClassComponent(renderBlockActions) || isFunctionComponent(renderBlockActions)) {
```

Comment:
Hm, I guess I don't comprehend this fully. Does that mean the `isClassComponent` should be named `isLegacyClassComponent`?

BTW, the `isValidElementType` returns true if a component is class-based component. But perhaps you mean it would return `true` for any function not just functional components?

Reply:
Yeah, seems like it returns true for even text (String).

Reply:
Yes, but it's not what we are looking for here.